pkg/kubelet/status/status_manager.go

```diff
@@ -282,6 +282,24 @@ func updatePodFromAllocation(pod *v1.Pod, allocs state.PodResourceAllocation) (*
 			}
 		}
 	}
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.SidecarContainers) {
+		for i, c := range pod.Spec.InitContainers {
+			if kubetypes.IsRestartableInitContainer(&c) {
+				if cAlloc, ok := allocated[c.Name]; ok {
+					if !apiequality.Semantic.DeepEqual(c.Resources, cAlloc) {
+						// Allocation differs from pod spec, update
+						if !updated {
+							// If this is the first update, copy the pod
+							pod = pod.DeepCopy()
+							updated = true
+						}
+						pod.Spec.InitContainers[i].Resources = cAlloc
+					}
+				}
```

Comment:
extract this inner part to a helper, `updateContainerFromAllocation`

Reply:
I tried to do it earlier as well but unit tests are failing. So I thought to not refactor the code as of now. 

Code: https://paste.googleplex.com/5491118887403520

Failed tests: https://paste.googleplex.com/4994718714036224

I guess this line: `if cAlloc, ok := allocated[c.Name]; ok ` is the culprit since cAlloc is the local variable and we are pointing `c.Resources` to it.

Reply:
How about this:
```
func updatePodFromAllocation(pod *v1.Pod, allocs state.PodResourceAllocation) (*v1.Pod, bool) {
	allocated, found := allocs[string(pod.UID)]
	if !found {
		return pod, false
	}

	updated := false
	updateContainerResources := func(containers []v1.Container, isInitContainer bool) []v1.Container {
		for i, c := range containers {
			if isInitContainer && !podutil.IsRestartableInitContainer(&c) {
				continue
			}

			if cAlloc, ok := allocated[c.Name]; ok {
				if !apiequality.Semantic.DeepEqual(c.Resources, cAlloc) {
					if !updated {
						pod = pod.DeepCopy()
						updated = true
					}
					containers[i].Resources = cAlloc
				}
			}
		}
		return containers
	}

	pod.Spec.Containers = updateContainerResources(pod.Spec.Containers, false)

	if utilfeature.DefaultFeatureGate.Enabled(features.SidecarContainers) {
		pod.Spec.InitContainers = updateContainerResources(pod.Spec.InitContainers, true)
	}

	return pod, updated
}
```

Reply:
This code will not work since you are just changing the local copy passed to the function (i.e `containers []v1.Container`) which will not be reflected in the actual pod spec (`pod.Spec.Containers.Resources`).

Reply:
Thanks for the code snippet. But unit tests are failing on my end when I tried to incorporate your snippet. I guess `containers[i].Resources` is also a pointer, and `Pod.Spec.Containers.Resources` is not getting updated even when updating `containers []v1.Container.Resources` (which is passed as reference by copy to the function) in the function.

Reply:
I applied this change locally and ran the unit tests and everything worked fine. Can you share the details of why the test failed?

<details>

<summary>diff --git a/pkg/kubelet/status/status_manager.go b/pkg/kubelet/status/status_manager.go</summary>


```diff
@@ -269,37 +269,31 @@ func updatePodFromAllocation(pod *v1.Pod, allocs state.PodResourceAllocation) (*
        }

        updated := false
-       for i, c := range pod.Spec.Containers {
-               if cAlloc, ok := allocated[c.Name]; ok {
-                       if !apiequality.Semantic.DeepEqual(c.Resources, cAlloc) {
-                               // Allocation differs from pod spec, update
-                               if !updated {
-                                       // If this is the first update, copy the pod
-                                       pod = pod.DeepCopy()
-                                       updated = true
-                               }
-                               pod.Spec.Containers[i].Resources = cAlloc
+       updateContainerResources := func(containers []v1.Container, isInitContainer bool) []v1.Container {
+               for i, c := range containers {
+                       if isInitContainer && !podutil.IsRestartableInitContainer(&c) {
+                               continue
                        }
-               }
-       }

-       if utilfeature.DefaultFeatureGate.Enabled(features.SidecarContainers) {
-               for i, c := range pod.Spec.InitContainers {
-                       if podutil.IsRestartableInitContainer(&c) {
-                               if cAlloc, ok := allocated[c.Name]; ok {
-                                       if !apiequality.Semantic.DeepEqual(c.Resources, cAlloc) {
-                                               // Allocation differs from pod spec, update
-                                               if !updated {
-                                                       // If this is the first update, copy the pod
-                                                       pod = pod.DeepCopy()
-                                                       updated = true
-                                               }
-                                               pod.Spec.InitContainers[i].Resources = cAlloc
+                       if cAlloc, ok := allocated[c.Name]; ok {
+                               if !apiequality.Semantic.DeepEqual(c.Resources, cAlloc) {
+                                       if !updated {
+                                               pod = pod.DeepCopy()
+                                               updated = true
                                        }
+                                       containers[i].Resources = cAlloc
                                }
                        }
                }
+               return containers
        }
+
+       pod.Spec.Containers = updateContainerResources(pod.Spec.Containers, false)
+
+       if utilfeature.DefaultFeatureGate.Enabled(features.SidecarContainers) {
+               pod.Spec.InitContainers = updateContainerResources(pod.Spec.InitContainers, true)
+       }
+
        return pod, updated
 }
```

</details>

```
go test -run  TestUpdatePodFromAllocation -v
=== RUN   TestUpdatePodFromAllocation
=== RUN   TestUpdatePodFromAllocation/steady_state
=== RUN   TestUpdatePodFromAllocation/no_allocations
=== RUN   TestUpdatePodFromAllocation/missing_container_allocation
=== RUN   TestUpdatePodFromAllocation/resized_container
--- PASS: TestUpdatePodFromAllocation (0.00s)
    --- PASS: TestUpdatePodFromAllocation/steady_state (0.00s)
    --- PASS: TestUpdatePodFromAllocation/no_allocations (0.00s)
    --- PASS: TestUpdatePodFromAllocation/missing_container_allocation (0.00s)
    --- PASS: TestUpdatePodFromAllocation/resized_container (0.00s)
PASS
ok  	k8s.io/kubernetes/pkg/kubelet/status	0.011s
```